conda_forge_tick/git_utils.py

```diff
@@ -134,37 +168,68 @@ class GitCli:
     If this does impact performance too much, we can consider a per-repository locking strategy.
     """
 
-    @staticmethod
+    def __init__(self):
+        self.__hidden_tokens: list[str] = []
+
     @lock_git_operation()
     def _run_git_command(
+        self,
         cmd: list[str | Path],
         working_directory: Path | None = None,
         check_error: bool = True,
-        capture_text: bool = False,
     ) -> subprocess.CompletedProcess:
         """
-        Run a git command.
+        Run a git command. stdout is only printed if the command fails. stderr is printed by default.
+        stdout is always captured, we capture stderr only if tokens are hidden.
         :param cmd: The command to run, as a list of strings.
         :param working_directory: The directory to run the command in. If None, the command will be run in the current
         working directory.
         :param check_error: If True, raise a GitCliError if the git command fails.
-        :param capture_text: If True, capture the output of the git command as text. The output will be in the
-        returned result object.
         :return: The result of the git command.
         :raises GitCliError: If the git command fails and check_error is True.
         :raises FileNotFoundError: If the working directory does not exist.
         """
         git_command = ["git"] + cmd
 
         logger.debug(f"Running git command: {git_command}")
-        capture_args = {"capture_output": True, "text": True} if capture_text else {}
+
+        # we only need to capture stderr if we want to hide tokens
+        stderr_args = {"stderr": subprocess.PIPE} if self.__hidden_tokens else {}
 
         try:
-            return subprocess.run(
-                git_command, check=check_error, cwd=working_directory, **capture_args
+            p = subprocess.run(
+                git_command,
+                check=check_error,
+                cwd=working_directory,
+                stdout=subprocess.PIPE,
+                **stderr_args,
+                text=True,
             )
         except subprocess.CalledProcessError as e:
-            raise GitCliError("Error running git command.") from e
+            e.stdout = replace_tokens(e.stdout, self.__hidden_tokens)
```

Comment:
I'm not a huge fan of encoding the credentials in the remote URL and then censoring the logs. I think a better approach would be similar to how GitHub handles it in the checkout action, by setting up the `'http.https://github.com/.extraheader'` git config to include the authorization header.

Reply:
Yeah putting the credentials on disk is not so great too right, or did I miss something? In any case, I don't actually care either way, but if we do put credentials on disk, we'll want to ensure they get cleaned up before the python process exits. On CI, GitHub will do this for us but potentially people running locally with a token might not.

Reply:
The feedstock cloning runs within a tmpdir context anyway. So if per-repository tokens are set, they are cleaned up automatically already.
So my proposal would be to implement it as @0xbe7a suggested and remove the token censoring entirely.

Reply:
Sorry for the delay - this is now implemented. For now, I only set the token for pushing and clear it immediately afterwards as an additional defense-in-depth measure.